---
title: "chromomap F2-28"
output: pdf_document
---

```{r setup, include=FALSE, message=FALSE}
library("dplyr")
library("ggplot2")
library("tidyr")
library("magrittr")
#suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("chromoMap"))
suppressPackageStartupMessages(library("ape")) # to read the GFF annotation file
```

# Import chromosome file

```{r}
chroms <- read.delim("../info/SL2.50_chromosomes.txt", sep="\t", header = FALSE, stringsAsFactors = F)
```

# Create annotation file

## Import list of DE genes from F2-28

```{r}
deg_F2_28 <- read.csv("../tables_F2-28/diff_res_F2-28orPI127826_vs_Elite.csv", 
                      stringsAsFactors = F, 
                      check.names = F) %>%
  select(gene, log2ratio_P28, pval_P28) %>%
  filter(pval_P28 < 0.01)
head(deg_F2_28)
```

## Import complete annotation data

:warning: check that the GFF3 file comes from the same genome version e.g. ITAG2.4 and SL2.50.  
```{r}
gff = read.gff(file = "../info/ITAG2.4_gene_models.modified.gff3", GFF3 = TRUE) %>%
  filter(type == "mRNA") %>%
  select(seqid, start, end, attributes) 
gff_filtered <- mutate(.data = gff, 
              gene = substr(attributes, start = 4, stop = 21)) %>%
  select(- attributes)
rm(gff)
head(gff_filtered)  
```

## Add annotation info to list of DEG genes

```{r}
deg_F2_28_with_annots <- inner_join(x = deg_F2_28, y = gff_filtered, by = "gene") %>%
  relocate(seqid, start, end, .before = "log2ratio_P28") %>%
  select(- pval_P28)
head(deg_F2_28_with_annots)
```

```{r}
write.table(deg_F2_28_with_annots, file = "../chromomaps/F2_28_annotation.tsv", sep = "\t", row.names = F, quote = F)
```

# Create chromomap plot

```{r}
png("../chromomaps/F2_28_chromomap.png", width = 12, height = 8)
chromoMap(ch.files = list(chroms), 
          data.files = list(deg_F2_28_with_annots), 
          data_based_color_map = TRUE, 
          data_type = "numeric", 
          legend = TRUE, lg_x = 20, lg_y = 200)
dev.off()
```



